### ISSUE TITLE
Improve podcast fetching

### ISSUE THREAD

**Issue by lightalloy** [thepracticaldev/dev.to#2952]
This is a part of Podcast section enrichment #271 

- sometimes "This podcast may not be playable" is displayed when a podcast is playable (https://dev.to/jsjabber/mjs-107-dan-fernandez), fix it
- find out why some podcasts are not playable and how to deal with it
- rescue specific errors instead of `StandardError`
- pull episodes asynchronously (1 job for each podcast)
- add more tests for `PodcastFeed` and refactor it

**Comment by Zhao-Andy** [comment]
Quick point about the "This podcast may not be playable": it was originally intended to display for podcasts that are hosted with SoundCloud. At the time, SoundCloud didn't have HTTPS links for their hosted podcast episodes. Non-HTTPS links were therefore not playable inside the `<audio>` tag. This might have changed now.

**Comment by lightalloy** [comment]
Thanks for the clarification, Andy. I'll consider this.

**Comment by benhalpern** [comment]
Yeah, it also acted as a bit of a catch-all for problems we were having.

My thoughts are that _if_ a podcast isn't playable, aka we detect as such on the client, we send a ping to the server to mark it as unplayable and then have an admin dashboard where we investigate.

If any podcast is wholly unplayable due to https-related issues, we should just delist it. This is not really related to fetching but I thought I'd dip in to provide some ideas.

Great stuff getting this under way! It will be great to have this section of the site be something we're proud of.

**Comment by lightalloy** [comment]
@Zhao-Andy I made a bit of research on the podcasts marked as unplayable. It seems like the current unplayable status is not related to SoundCloud anymore.
Podcasts are marked as unplayable if one of the episodes failed to be fetched from its url.
https://github.com/thepracticaldev/dev.to/blob/8c52d5c43f1d38d5d71453f99a5292d6d7e582ec/app/labor/podcast_feed.rb#L65-L76

I think it makes more sense to mark an episode nonplayable, not a podcast. I agree with @ben on delisting and displaying on admin dashboard to investigate, though I'm not sure how it should look like for admins. I'll start with adding tests and refactoring to understand the feature better.

**Comment by lightalloy** [comment]
Some thoughts on the podcasts playability after more research:

- as Andy said, non-https urls are not allowed in the `<audio>` tag
- some podcasts don't provide `https` media urls, but sometimes the episode is actually reachable by `https`, though there is an `http` url in the feed

So, currently, if a podcast episode has an `http` url in the feed, we try to reach it by `https`. If it's reachable by https, we set the episode url to `https`, if not, we set the "This podcast may not be playable in the browser" status notice.
Previously we discussed that we should delist unplayable podcasts from the site, but having potentially unplayable podcasts makes sense because even if the episode is unplayable, the user can still download it and listen to the podcast.

Another thing that I noticed is that some of the unplayable episodes (with http media url) are reachable by `https`, so if we replace `http` with `https` they will be playable. Maybe they were marked as unplayable because of temporary unavailability.

What I suggest:
- distinguish unreachable and potentially unplayable episodes.
If an episode is unreachable, we should delist it, if it's just unplayable, it still makes sense to keep it.
- set statuses for episodes, not for the whole podcasts. This way we can be more precise about the episodes' statuses.
- I think it makes sense to update urls for the existing "unplayable" episodes (try to reach the `https` urls and update `media_url`'s if they are reachable)
- probably, it makes sense to try to reach an unplayable/unreachable episode several times. Currently, it's only tried to reach one time and marked as unplayable. The url is only updated only if it has changed in the feed to the `https` version.
